diff --git a/avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/JdbcMeta.java b/avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/JdbcMeta.java
index 276dd71b5..3b241de6e 100644
--- a/avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/JdbcMeta.java
+++ b/avatica-server/src/main/java/org/apache/calcite/avatica/jdbc/JdbcMeta.java
@@ -184,6 +184,9 @@ public String defaultValue() {
    */
   protected static List<ColumnMetaData>
   columns(ResultSetMetaData metaData) throws SQLException {
+    if (metaData == null) {
+      return Collections.emptyList();
+    }
     final List<ColumnMetaData> columns = new ArrayList<>();
     for (int i = 1; i <= metaData.getColumnCount(); i++) {
       final Type javaType =
@@ -805,10 +808,15 @@ public Frame fetch(StatementHandle h, List<Object> parameterValues,
             preparedStatement.setObject(i + 1, o);
           }
         }
-        statementInfo.resultSet = preparedStatement.executeQuery();
+        if (preparedStatement.execute()) {
+          statementInfo.resultSet = preparedStatement.getResultSet();
+        }
+      }
+      if (statementInfo.resultSet == null) {
+        return Frame.EMPTY;
+      } else {
+        return JdbcResultSet.frame(statementInfo.resultSet, offset, fetchMaxRowCount);
       }
-      return JdbcResultSet.frame(statementInfo.resultSet, offset,
-          fetchMaxRowCount);
     } catch (SQLException e) {
       throw propagate(e);
     }
diff --git a/avatica-server/src/test/java/org/apache/calcite/avatica/RemoteDriverTest.java b/avatica-server/src/test/java/org/apache/calcite/avatica/RemoteDriverTest.java
index f23e2fcae..1c0e7d91c 100644
--- a/avatica-server/src/test/java/org/apache/calcite/avatica/RemoteDriverTest.java
+++ b/avatica-server/src/test/java/org/apache/calcite/avatica/RemoteDriverTest.java
@@ -195,6 +195,48 @@ private void checkStatementExecuteQuery(Connection connection,
     connection.close();
   }
 
+  @Test public void testStatementExecuteLocal() throws Exception {
+    checkStatementExecute(ljs(), false);
+  }
+
+  @Test public void testPrepareExecuteLocal() throws Exception {
+    checkStatementExecute(ljs(), true);
+  }
+
+  private void checkStatementExecute(Connection connection, boolean prepare) throws SQLException {
+    final String sql = "select * from (\n"
+        + "  values (1, 'a'), (null, 'b'), (3, 'c')) as t (c1, c2)";
+    final Statement statement;
+    final ResultSet resultSet;
+    final ParameterMetaData parameterMetaData;
+    if (prepare) {
+      final PreparedStatement ps = connection.prepareStatement(sql);
+      statement = ps;
+      parameterMetaData = ps.getParameterMetaData();
+      assertTrue(ps.execute());
+      resultSet = ps.getResultSet();
+    } else {
+      statement = connection.createStatement();
+      parameterMetaData = null;
+      assertTrue(statement.execute(sql));
+      resultSet = statement.getResultSet();
+    }
+    if (parameterMetaData != null) {
+      assertThat(parameterMetaData.getParameterCount(), equalTo(0));
+    }
+    final ResultSetMetaData metaData = resultSet.getMetaData();
+    assertEquals(2, metaData.getColumnCount());
+    assertEquals("C1", metaData.getColumnName(1));
+    assertEquals("C2", metaData.getColumnName(2));
+    assertTrue(resultSet.next());
+    assertTrue(resultSet.next());
+    assertTrue(resultSet.next());
+    assertFalse(resultSet.next());
+    resultSet.close();
+    statement.close();
+    connection.close();
+  }
+
   @Test public void testCreateInsertUpdateDrop() throws Exception {
     final String drop = "drop table TEST_TABLE if exists";
     final String create = "create table TEST_TABLE("
diff --git a/avatica/src/main/java/org/apache/calcite/avatica/AvaticaPreparedStatement.java b/avatica/src/main/java/org/apache/calcite/avatica/AvaticaPreparedStatement.java
index 9324db7e4..f85dcc186 100644
--- a/avatica/src/main/java/org/apache/calcite/avatica/AvaticaPreparedStatement.java
+++ b/avatica/src/main/java/org/apache/calcite/avatica/AvaticaPreparedStatement.java
@@ -90,7 +90,8 @@ public ParameterMetaData getParameterMetaData() throws SQLException {
   }
 
   public int executeUpdate() throws SQLException {
-    throw new UnsupportedOperationException(); // TODO:
+    getConnection().executeQueryInternal(this, signature, null);
+    return updateCount;
   }
 
   public void setNull(int parameterIndex, int sqlType) throws SQLException {
@@ -186,7 +187,10 @@ public void setObject(int parameterIndex, Object x) throws SQLException {
   }
 
   public boolean execute() throws SQLException {
-    throw new UnsupportedOperationException();
+    getConnection().executeQueryInternal(this, signature, null);
+    // Result set is null for DML or DDL.
+    // Result set is closed if user cancelled the query.
+    return openResultSet != null && !openResultSet.isClosed();
   }
 
   public void addBatch() throws SQLException {
