diff --git a/avatica/server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java b/avatica/server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java
index b047137d1..3050f726e 100644
--- a/avatica/server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java
+++ b/avatica/server/src/main/java/org/apache/calcite/avatica/server/HttpServer.java
@@ -242,8 +242,24 @@ protected ConstraintSecurityHandler configureSpnego(Server server, ServerConnect
     PropertyBasedSpnegoLoginService spnegoLoginService =
         new PropertyBasedSpnegoLoginService(realm, principal);
 
+    // Roles are "realms" for Kerberos/SPNEGO
+    final String[] allowedRealms = getAllowedRealms(realm, config);
+
     return configureCommonAuthentication(server, connector, config, Constraint.__SPNEGO_AUTH,
-        new String[] {realm}, new SpnegoAuthenticator(), realm, spnegoLoginService);
+        allowedRealms, new SpnegoAuthenticator(), realm, spnegoLoginService);
+  }
+
+  protected String[] getAllowedRealms(String serverRealm, AvaticaServerConfiguration config) {
+    // Roles are "realms" for Kerberos/SPNEGO
+    String[] allowedRealms = new String[] {serverRealm};
+    // By default, only the server's realm is allowed, but other realms can also be allowed.
+    if (null != config.getAllowedRoles()) {
+      allowedRealms = new String[config.getAllowedRoles().length + 1];
+      allowedRealms[0] = serverRealm;
+      System.arraycopy(config.getAllowedRoles(), 0, allowedRealms, 1,
+          config.getAllowedRoles().length);
+    }
+    return allowedRealms;
   }
 
   protected ConstraintSecurityHandler configureBasicAuthentication(Server server,
@@ -315,6 +331,10 @@ protected ServerConnector configureConnector(ServerConnector connector, int port
     return connector;
   }
 
+  protected AvaticaServerConfiguration getConfig() {
+    return this.config;
+  }
+
   public void stop() {
     if (server == null) {
       throw new RuntimeException("Server is already stopped");
@@ -416,13 +436,29 @@ public Builder withMetricsConfiguration(MetricsSystemConfiguration<?> metricsCon
      * @return <code>this</code>
      */
     public Builder withSpnego(String principal) {
+      return withSpnego(principal, (String[]) null);
+    }
+
+    /**
+     * Configures the server to use SPNEGO authentication. This method requires that the
+     * <code>principal</code> contains the Kerberos realm. Invoking this method overrides any
+     * previous call which configures authentication. Invoking this method overrides any previous
+     * call which configures authentication. By default, only principals from the server's realm are
+     * permitted, but additional realms can be allowed using <code>additionalAllowedRealms</code>.
+     *
+     * @param principal A kerberos principal with the realm required.
+     * @param additionalAllowedRealms Any additional realms, other than the server's realm, which
+     *    should be allowed to authenticate against the server. Can be null.
+     * @return <code>this</code>
+     */
+    public Builder withSpnego(String principal, String[] additionalAllowedRealms) {
       int index = Objects.requireNonNull(principal).lastIndexOf('@');
       if (-1 == index) {
         throw new IllegalArgumentException("Could not find '@' symbol in '" + principal
             + "' to parse the Kerberos realm from the principal");
       }
       final String realm = principal.substring(index + 1);
-      return withSpnego(principal, realm);
+      return withSpnego(principal, realm, additionalAllowedRealms);
     }
 
     /**
@@ -437,9 +473,28 @@ public Builder withSpnego(String principal) {
      * @return <code>this</code>
      */
     public Builder withSpnego(String principal, String realm) {
+      return this.withSpnego(principal, realm, null);
+    }
+
+    /**
+     * Configures the server to use SPNEGO authentication. It is required that callers are logged
+     * in via Kerberos already or have provided the necessary configuration to automatically log
+     * in via JAAS (using the <code>java.security.auth.login.config</code> system property) before
+     * starting the {@link HttpServer}. Invoking this method overrides any previous call which
+     * configures authentication. By default, only principals from the server's realm are permitted,
+     * but additional realms can be allowed using <code>additionalAllowedRealms</code>.
+     *
+     * @param principal The kerberos principal
+     * @param realm The kerberos realm
+     * @param additionalAllowedRealms Any additional realms, other than the server's realm, which
+     *    should be allowed to authenticate against the server. Can be null.
+     * @return <code>this</code>
+     */
+    public Builder withSpnego(String principal, String realm, String[] additionalAllowedRealms) {
       this.authenticationType = AuthenticationType.SPNEGO;
       this.kerberosPrincipal = Objects.requireNonNull(principal);
       this.kerberosRealm = Objects.requireNonNull(realm);
+      this.loginServiceAllowedRoles = additionalAllowedRealms;
       return this;
     }
 
@@ -568,6 +623,7 @@ private AvaticaHandler buildHandler(Builder b, AvaticaServerConfiguration config
     private AvaticaServerConfiguration buildSpnegoConfiguration(Builder b) {
       final String principal = b.kerberosPrincipal;
       final String realm = b.kerberosRealm;
+      final String[] additionalAllowedRealms = b.loginServiceAllowedRoles;
       final DoAsRemoteUserCallback callback = b.remoteUserCallback;
       return new AvaticaServerConfiguration() {
 
@@ -593,7 +649,7 @@ private AvaticaServerConfiguration buildSpnegoConfiguration(Builder b) {
         }
 
         @Override public String[] getAllowedRoles() {
-          return null;
+          return additionalAllowedRealms;
         }
 
         @Override public String getHashLoginServiceRealm() {
