diff --git a/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java b/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
index 52a1cfc46..903332211 100644
--- a/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
+++ b/core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
@@ -3458,22 +3458,21 @@ private RexNode convertUsing(SqlValidatorNamespace leftNamespace,
           validator().getTypeCoercion().commonTypeForBinaryComparison(
               comparedTypes.get(0), comparedTypes.get(1));
       if (resultType == null) {
-        // This should never happen, since the program has been validated.
-        throw new IllegalArgumentException("Cannot join on field `" + name
-            + "` because the types are not comparable: " + comparedTypes);
-      }
-
-      List<RexNode> castedOperands = new ArrayList<>();
-      for (int i = 0; i < operands.size(); i++) {
-        RexNode operand = operands.get(i);
-        RelDataType fieldType = comparedTypes.get(i);
-        RexNode expr = operand;
-        if (!fieldType.equals(resultType)) {
-          expr = rexBuilder.makeCast(resultType, operand, true, false);
+        // Leave call unchanged (as it happens in TypeCoercionImpl#binaryComparisonCoercion)
+        list.add(rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, operands));
+      } else {
+        List<RexNode> castedOperands = new ArrayList<>();
+        for (int i = 0; i < operands.size(); i++) {
+          RexNode operand = operands.get(i);
+          RelDataType fieldType = comparedTypes.get(i);
+          RexNode expr = operand;
+          if (!fieldType.equals(resultType)) {
+            expr = rexBuilder.makeCast(resultType, operand, true, false);
+          }
+          castedOperands.add(expr);
         }
-        castedOperands.add(expr);
+        list.add(rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, castedOperands));
       }
-      list.add(rexBuilder.makeCall(SqlStdOperatorTable.EQUALS, castedOperands));
     }
     return RexUtil.composeConjunction(rexBuilder, list);
   }
